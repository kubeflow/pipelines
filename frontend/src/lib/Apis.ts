// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as portableFetch from 'portable-fetch';
import { ExperimentServiceApi, FetchAPI } from '../apis/experiment';
import { JobServiceApi } from '../apis/job';
import { ApiPipeline, ApiPipelineVersion, PipelineServiceApi } from '../apis/pipeline';
import { RunServiceApi } from '../apis/run';
import { ApiVisualization, VisualizationServiceApi } from '../apis/visualization';
import { HTMLViewerConfig } from '../components/viewers/HTMLViewer';
import { PlotType } from '../components/viewers/Viewer';
import * as Utils from './Utils';
import { StoragePath } from './WorkflowParser';
import { buildQuery } from './Utils';

const v1beta1Prefix = 'apis/v1beta1';

export interface ListRequest {
  filter?: string;
  orderAscending?: boolean;
  pageSize?: number;
  pageToken?: string;
  sortBy?: string;
}

export interface BuildInfo {
  apiServerCommitHash?: string;
  apiServerTagName?: string;
  apiServerReady?: boolean;
  buildDate?: string;
  frontendCommitHash?: string;
  frontendTagName?: string;
}

// Hack types from https://github.com/microsoft/TypeScript/issues/1897#issuecomment-557057387
export type JSONPrimitive = string | number | boolean | null;
export type JSONValue = JSONPrimitive | JSONObject | JSONArray;
export type JSONObject = { [member: string]: JSONValue };
export type JSONArray = JSONValue[];

let customVisualizationsAllowed: boolean;

// For cross browser support, fetch should use 'same-origin' as default. This fixes firefox auth issues.
// Refrence: https://github.com/github/fetch#sending-cookies
const crossBrowserFetch: FetchAPI = (url, init) =>
  portableFetch(url, { credentials: 'same-origin', ...init });

export class Apis {
  public static async areCustomVisualizationsAllowed(): Promise<boolean> {
    // Result is cached to prevent excessive network calls for simple request.
    // The value of customVisualizationsAllowed will only change if the
    // deployment is updated and then the entire pod is restarted.
    if (customVisualizationsAllowed === undefined) {
      const result = await this._fetch('visualizations/allowed');
      customVisualizationsAllowed = result === 'true';
    }
    return customVisualizationsAllowed;
  }

  public static async buildPythonVisualizationConfig(
    visualizationData: ApiVisualization,
    namespace?: string,
  ): Promise<HTMLViewerConfig> {
    const visualization = await Apis.visualizationServiceApi.createVisualization(
      namespace || '',
      visualizationData,
    );
    if (visualization.html) {
      const htmlContent = visualization.html
        // Fixes issue with TFX components (and other iframe based
        // visualizations), where the method in which javascript interacts
        // with embedded iframes is not allowed when embedded in an additional
        // iframe. This is resolved by setting the srcdoc value rather that
        // manipulating the document directly.
        .replace('contentWindow.document.write', 'srcdoc=');
      return {
        htmlContent,
        type: PlotType.WEB_APP,
      } as HTMLViewerConfig;
    } else {
      // This should never be thrown as the html property of a generated
      // visualization is always set for successful visualization generations.
      throw new Error('Visualization was generated successfully but generated HTML was not found.');
    }
  }

  /**
   * Get pod logs
   */
  public static getPodLogs(runId: string, podName: string, podNamespace: string): Promise<string> {
    let query = `k8s/pod/logs?podname=${encodeURIComponent(podName)}&runid=${encodeURIComponent(
      runId,
    )}`;
    if (podNamespace) {
      query += `&podnamespace=${encodeURIComponent(podNamespace)}`;
    }
    return this._fetch(query);
  }

  /**
   * Get pod info
   */
  public static async getPodInfo(podName: string, podNamespace: string): Promise<JSONObject> {
    const query = `k8s/pod?podname=${encodeURIComponent(podName)}&podnamespace=${encodeURIComponent(
      podNamespace,
    )}`;
    const podInfo = await this._fetch(query);
    return JSON.parse(podInfo);
  }

  /**
   * Get pod events
   */
  public static async getPodEvents(podName: string, podNamespace: string): Promise<JSONObject> {
    const query = `k8s/pod/events?podname=${encodeURIComponent(
      podName,
    )}&podnamespace=${encodeURIComponent(podNamespace)}`;
    const eventList = await this._fetch(query);
    return JSON.parse(eventList);
  }

  public static get basePath(): string {
    const path = window.location.protocol + '//' + window.location.host + window.location.pathname;
    // Trim trailing '/' if exists
    return path.endsWith('/') ? path.substr(0, path.length - 1) : path;
  }

  public static get experimentServiceApi(): ExperimentServiceApi {
    if (!this._experimentServiceApi) {
      this._experimentServiceApi = new ExperimentServiceApi(
        { basePath: this.basePath },
        undefined,
        crossBrowserFetch,
      );
    }
    return this._experimentServiceApi;
  }

  public static get jobServiceApi(): JobServiceApi {
    if (!this._jobServiceApi) {
      this._jobServiceApi = new JobServiceApi(
        { basePath: this.basePath },
        undefined,
        crossBrowserFetch,
      );
    }
    return this._jobServiceApi;
  }

  public static get pipelineServiceApi(): PipelineServiceApi {
    if (!this._pipelineServiceApi) {
      this._pipelineServiceApi = new PipelineServiceApi(
        { basePath: this.basePath },
        undefined,
        crossBrowserFetch,
      );
    }
    return this._pipelineServiceApi;
  }

  public static get runServiceApi(): RunServiceApi {
    if (!this._runServiceApi) {
      this._runServiceApi = new RunServiceApi(
        { basePath: this.basePath },
        undefined,
        crossBrowserFetch,
      );
    }
    return this._runServiceApi;
  }

  public static get visualizationServiceApi(): VisualizationServiceApi {
    if (!this._visualizationServiceApi) {
      this._visualizationServiceApi = new VisualizationServiceApi(
        { basePath: this.basePath },
        undefined,
        crossBrowserFetch,
      );
    }
    return this._visualizationServiceApi;
  }

  /**
   * Retrieve various information about the build.
   */
  public static async getBuildInfo(): Promise<BuildInfo> {
    return await this._fetchAndParse<BuildInfo>('/healthz', v1beta1Prefix);
  }

  /**
   * Verifies that Jupyter Hub is reachable.
   */
  public static async isJupyterHubAvailable(): Promise<boolean> {
    const response = await fetch('/hub/', { credentials: 'same-origin' });
    return response ? response.ok : false;
  }

  /**
   * Reads file from storage using server.
   */
  public static readFile(path: StoragePath, namespace?: string, peek?: number): Promise<string> {
    return this._fetch(this.buildReadFileUrl({ path, namespace, peek, isDownload: false }));
  }

  /**
   * Builds an url for the readFile API to retrieve a workflow artifact.
   * @param path object describing the artifact (e.g. source, bucket, and key)
   * @param isDownload whether we download the artifact as is (e.g. skip extracting from *.tar.gz)
   */
  public static buildReadFileUrl({
    path,
    namespace,
    peek,
    isDownload,
  }: {
    path: StoragePath;
    namespace?: string;
    peek?: number;
    isDownload?: boolean;
  }) {
    const { source, bucket, key } = path;
    if (isDownload) {
      return `artifacts/${source}/${bucket}/${key}${buildQuery({
        namespace,
        peek,
      })}`;
    } else {
      return `artifacts/get${buildQuery({ source, namespace, peek, bucket, key })}`;
    }
  }

  /**
   * Builds an url to visually represents a workflow artifact location.
   * @param param.source source of the artifact (e.g. minio, gcs, s3, http, or https)
   * @param param.bucket name of the bucket with the artifact (or host for http/https)
   * @param param.key key (i.e. path) of the artifact in the bucket
   */
  public static buildArtifactUrl({ source, bucket, key }: StoragePath) {
    // TODO see https://github.com/kubeflow/pipelines/pull/3725
    return `${source}://${bucket}/${key}`;
  }

  /**
   * Gets the address (IP + port) of a Tensorboard service given the logdir and tfversion
   */
  public static getTensorboardApp(
    logdir: string,
    namespace: string,
  ): Promise<{ podAddress: string; tfVersion: string }> {
    return this._fetchAndParse<{ podAddress: string; tfVersion: string }>(
      `apps/tensorboard?logdir=${encodeURIComponent(logdir)}&namespace=${encodeURIComponent(
        namespace,
      )}`,
    );
  }

  /**
   * Starts a deployment and service for Tensorboard given the logdir
   */
  public static startTensorboardApp(
    logdir: string,
    tfversion: string,
    namespace: string,
  ): Promise<string> {
    return this._fetch(
      `apps/tensorboard?logdir=${encodeURIComponent(logdir)}&tfversion=${encodeURIComponent(
        tfversion,
      )}&namespace=${encodeURIComponent(namespace)}`,
      undefined,
      undefined,
      { headers: { 'content-type': 'application/json' }, method: 'POST' },
    );
  }

  /**
   * Check if the underlying Tensorboard pod is actually up, given the pod address
   */
  public static async isTensorboardPodReady(path: string): Promise<boolean> {
    const resp = await fetch(path, { method: 'HEAD' });
    return resp.ok;
  }

  /**
   * Delete a deployment and its service of the Tensorboard given the URL
   */
  public static deleteTensorboardApp(logdir: string, namespace: string): Promise<string> {
    return this._fetch(
      `apps/tensorboard?logdir=${encodeURIComponent(logdir)}&namespace=${encodeURIComponent(
        namespace,
      )}`,
      undefined,
      undefined,
      { method: 'DELETE' },
    );
  }

  /**
   * Uploads the given pipeline file to the backend, and gets back a Pipeline
   * object with its metadata parsed.
   */
  public static async uploadPipeline(
    pipelineName: string,
    pipelineDescription: string,
    pipelineData: File,
  ): Promise<ApiPipeline> {
    const fd = new FormData();
    fd.append('uploadfile', pipelineData, pipelineData.name);
    return await this._fetchAndParse<ApiPipeline>(
      '/pipelines/upload',
      v1beta1Prefix,
      `name=${encodeURIComponent(pipelineName)}&description=${encodeURIComponent(
        pipelineDescription,
      )}`,
      {
        body: fd,
        cache: 'no-cache',
        method: 'POST',
      },
    );
  }

  public static async uploadPipelineVersion(
    versionName: string,
    pipelineId: string,
    versionData: File,
  ): Promise<ApiPipelineVersion> {
    const fd = new FormData();
    fd.append('uploadfile', versionData, versionData.name);
    return await this._fetchAndParse<ApiPipelineVersion>(
      '/pipelines/upload_version',
      v1beta1Prefix,
      `name=${encodeURIComponent(versionName)}&pipelineid=${encodeURIComponent(pipelineId)}`,
      {
        body: fd,
        cache: 'no-cache',
        method: 'POST',
      },
    );
  }

  /*
   * Retrieves the name of the Kubernetes cluster if it is running in GKE, otherwise returns an error.
   */
  public static async getClusterName(): Promise<string> {
    return this._fetch('system/cluster-name');
  }

  /*
   * Retrieves the project ID in which this cluster is running if using GKE, otherwise returns an error.
   */
  public static async getProjectId(): Promise<string> {
    return this._fetch('system/project-id');
  }

  private static _experimentServiceApi?: ExperimentServiceApi;
  private static _jobServiceApi?: JobServiceApi;
  private static _pipelineServiceApi?: PipelineServiceApi;
  private static _runServiceApi?: RunServiceApi;
  private static _visualizationServiceApi?: VisualizationServiceApi;

  /**
   * This function will call this._fetch() and parse the resulting JSON into an object of type T.
   */
  private static async _fetchAndParse<T>(
    path: string,
    apisPrefix?: string,
    query?: string,
    init?: RequestInit,
  ): Promise<T> {
    const responseText = await this._fetch(path, apisPrefix, query, init);
    try {
      return JSON.parse(responseText) as T;
    } catch (err) {
      throw new Error(
        `Error parsing response for path: ${path}\n\n` +
          `Response was: ${responseText}\n\nError was: ${JSON.stringify(err)}`,
      );
    }
  }

  /**
   * Makes an HTTP request and returns the response as a string.
   *
   * Use this._fetchAndParse() if you intend to parse the response as JSON into an object.
   */
  private static async _fetch(
    path: string,
    apisPrefix?: string,
    query?: string,
    init?: RequestInit,
  ): Promise<string> {
    init = Object.assign(init || {}, { credentials: 'same-origin' });
    const response = await fetch((apisPrefix || '') + path + (query ? '?' + query : ''), init);
    const responseText = await response.text();
    if (response.ok) {
      return responseText;
    } else {
      Utils.logger.error(
        `Response for path: ${path} was not 'ok'\n\nResponse was: ${responseText}`,
      );
      throw new Error(responseText);
    }
  }
}

// Valid sortKeys as specified by the backend.
// Note that '' and 'created_at' are considered equivalent.
export enum RunSortKeys {
  CREATED_AT = 'created_at',
  NAME = 'name',
}

// Valid sortKeys as specified by the backend.
export enum PipelineSortKeys {
  AUTHOR = 'id',
  CREATED_AT = 'created_at',
  ID = 'id',
  NAME = 'name',
}

// Valid sortKeys as specified by the backend.
export enum JobSortKeys {
  CREATED_AT = 'created_at',
  ID = 'id',
  NAME = 'name',
  PIPELINE_ID = 'pipeline_id',
}

// Valid sortKeys as specified by the backend.
export enum ExperimentSortKeys {
  CREATED_AT = 'created_at',
  ID = 'id',
  NAME = 'name',
}

// Valid sortKeys as specified by the backend.
export enum PipelineVersionSortKeys {
  CREATED_AT = 'created_at',
  NAME = 'name',
}
