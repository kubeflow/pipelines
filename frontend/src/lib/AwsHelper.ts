/*
 * Copyright 2018-2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Check if the provided string is an S3 endpoint (can be any region).
 *
 * @param endpoint minio endpoint to check.
 */
export function isS3Endpoint(endpoint: string = ''): boolean {
  return !!endpoint.match(/s3.{0,}\.amazonaws\.com\.?.{0,}/i);
}
