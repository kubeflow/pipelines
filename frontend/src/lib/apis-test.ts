import * as Apis from './apis';
import * as sinon from 'sinon';

const testTemplates = [{
  "id": 1,
  "name": "Unstructured text",
  "author": "user1",
  "tags": [
    "text",
    "experiment"
  ],
  "description": "An awesome unstructured text pipeline template.",
  "location": "gcs://pipelines_bucket/structured_text1",
  "parameters": ["x", "y"],
  "sharedWith": "team1"
}];

describe('My API', () => {
  let server: sinon.SinonFakeServer;

  beforeEach(() => {
    server = sinon.fakeServer.create();
  });

  afterEach(() => {
    server.restore();
  });

  it('should request _templates', async (done) => {
    server.respondWith('GET', '_templates',
      [200, { 'Content-Type': 'application/json' },
        JSON.stringify(testTemplates)]);

    try {
      const response = await Apis.getTemplates();
      assert.equal(response, JSON.stringify(testTemplates));
      done();
    } catch (err) {
      done(err);
    }
    server.respond();
  });
});