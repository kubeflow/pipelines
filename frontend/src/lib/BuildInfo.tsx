/**
 * Copyright 2023 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Apis, BuildInfo } from '../lib/Apis';
import { logger } from './Utils';

interface BuildInfoProviderState {
  buildInfo: BuildInfo | undefined;
}

export const BuildInfoContext = React.createContext<BuildInfo | undefined>(undefined);
export class BuildInfoProvider extends React.Component<{}, BuildInfoProviderState> {
  state = {
    buildInfo: undefined,
  };
  async componentDidMount() {
    try {
      const buildInfo = await Apis.getBuildInfo();
      this.setState({ buildInfo });
    } catch (err) {
      logger.error('Failed to retrieve build info', err);
    }
  }
  render() {
    return <BuildInfoContext.Provider value={this.state.buildInfo} {...this.props} />;
  }
}
