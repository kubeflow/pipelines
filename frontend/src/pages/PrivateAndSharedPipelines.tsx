/*
 * Copyright 2023 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { classes } from 'typestyle';
import MD2Tabs from '../atoms/MD2Tabs';
import { PageProps } from './Page';
import PipelineList from './PipelineList';
import { RoutePage } from '../components/Router';
import { NamespaceContext } from '../lib/KubeflowClient';
import { commonCss, padding } from '../Css';
import { BuildInfoContext } from 'src/lib/BuildInfo';

export enum PrivateAndSharedTab {
  PRIVATE = 0,
  SHARED = 1,
}

export interface PrivateAndSharedProps extends PageProps {
  view: PrivateAndSharedTab;
}

const PrivatePipelineList: React.FC<PageProps> = props => {
  const namespace = React.useContext(NamespaceContext);
  return <PipelineList {...props} namespace={namespace} />;
};

export enum PipelineTabsHeaders {
  PRIVATE = 'Private',
  SHARED = 'Shared',
}

export enum PipelineTabsTooltips {
  PRIVATE = 'Only people who have access to this namespace will be able to view and use these pipelines.',
  SHARED = 'Everyone in your organization will be able to view and use these pipelines.',
}

const PrivateAndSharedPipelines: React.FC<PrivateAndSharedProps> = props => {
  const buildInfo = React.useContext(BuildInfoContext);

  const tabSwitched = (newTab: PrivateAndSharedTab): void => {
    props.history.push(
      newTab === PrivateAndSharedTab.PRIVATE ? RoutePage.PIPELINES : RoutePage.PIPELINES_SHARED,
    );
  };

  if (!buildInfo?.apiServerMultiUser) {
    return <PipelineList {...props} />;
  }
  return (
    <div className={classes(commonCss.page, padding(20, 't'))}>
      <MD2Tabs
        tabs={[
          {
            header: PipelineTabsHeaders.PRIVATE,
            tooltip: PipelineTabsTooltips.PRIVATE,
          },
          {
            header: PipelineTabsHeaders.SHARED,
            tooltip: PipelineTabsTooltips.SHARED,
          },
        ]}
        selectedTab={props.view}
        onSwitch={tabSwitched}
      />

      {props.view === PrivateAndSharedTab.PRIVATE && <PrivatePipelineList {...props} />}

      {props.view === PrivateAndSharedTab.SHARED && <PipelineList {...props} />}
    </div>
  );
};

export default PrivateAndSharedPipelines;
