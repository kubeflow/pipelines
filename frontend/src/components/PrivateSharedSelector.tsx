/*
 * Copyright 2023 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { classes } from 'typestyle';
import Radio from '@material-ui/core/Radio';
import Tooltip from '@material-ui/core/Tooltip';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { commonCss, padding } from '../Css';
import { PipelineTabsHeaders } from '../pages/PrivateAndSharedPipelines';

export interface PrivateSharedSelectorProps {
  onChange: (isPrivate: boolean) => void;
}

export enum PipelineButtonTooltips {
  PRIVATE = 'Only people who have access to this namespace will be able to view and use this pipeline.',
  SHARED = 'Everyone in your organization will be able to view and use this pipeline.',
}

const PrivateSharedSelector: React.FC<PrivateSharedSelectorProps> = (props): JSX.Element | null => {
  const [namespacedPipeline, setNamespacedPipeline] = React.useState(true);

  React.useEffect(() => {
    props.onChange(namespacedPipeline);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [namespacedPipeline]);

  return (
    <React.Fragment>
      <div>Select if the new pipeline will be private or shared.</div>
      <div className={classes(commonCss.flex, padding(10, 'b'))}>
        <Tooltip title={PipelineButtonTooltips.PRIVATE} placement='top-start'>
          <FormControlLabel
            id='createNewPrivatePipelineBtn'
            label={PipelineTabsHeaders.PRIVATE}
            checked={namespacedPipeline === true}
            control={<Radio color='primary' />}
            onChange={() => {
              setNamespacedPipeline(true);
            }}
          />
        </Tooltip>
        <Tooltip title={PipelineButtonTooltips.SHARED} placement='top-start'>
          <FormControlLabel
            id='createNewSharedPipelineBtn'
            label={PipelineTabsHeaders.SHARED}
            checked={namespacedPipeline === false}
            control={<Radio color='primary' />}
            onChange={() => {
              setNamespacedPipeline(false);
            }}
          />
        </Tooltip>
      </div>
    </React.Fragment>
  );
};

export default PrivateSharedSelector;
