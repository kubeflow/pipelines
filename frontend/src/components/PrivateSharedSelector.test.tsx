/*
 * Copyright 2023 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { BuildInfoContext } from '../lib/BuildInfo';
import PrivateSharedSelector, { PrivateSharedSelectorProps } from './PrivateSharedSelector';
import { PipelineTabsHeaders } from '../pages/PrivateAndSharedPipelines';

function generateProps(): PrivateSharedSelectorProps {
  return {
    onChange: jest.fn(),
  };
}

describe('PrivateSharedSelector', () => {
  it('it renders correctly', async () => {
    const tree = render(<PrivateSharedSelector {...generateProps()} />);
    expect(tree).toMatchSnapshot();
  });

  it('it changes checked input on click', async () => {
    render(
      <BuildInfoContext.Provider value={{ apiServerMultiUser: true }}>
        <PrivateSharedSelector {...generateProps()} />
      </BuildInfoContext.Provider>,
    );

    const privateInput = screen.getByLabelText(PipelineTabsHeaders.PRIVATE) as HTMLInputElement;
    const sharedInput = screen.getByLabelText(PipelineTabsHeaders.SHARED) as HTMLInputElement;
    expect(privateInput.checked).toBe(true);
    expect(sharedInput.checked).toBe(false);

    fireEvent.click(sharedInput);

    expect(privateInput.checked).toBe(false);
    expect(sharedInput.checked).toBe(true);
  });
});
