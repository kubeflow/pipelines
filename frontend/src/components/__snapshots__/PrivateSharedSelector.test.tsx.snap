// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateSharedSelector it renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        Select if the new pipeline will be private or shared.
      </div>
      <div
        class="flex"
      >
        <label
          class="MuiFormControlLabel-root-9"
          id="createNewPrivatePipelineBtn"
          title="Only people who have access to this namespace will be able to view and use this pipeline."
        >
          <span
            class="MuiButtonBase-root-30 MuiIconButton-root-24 MuiPrivateSwitchBase-root-20 MuiRadio-root-15 MuiRadio-colorPrimary-18 MuiPrivateSwitchBase-checked-21 MuiRadio-checked-16"
          >
            <span
              class="MuiIconButton-label-29"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-33"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <input
                checked=""
                class="MuiPrivateSwitchBase-input-23"
                type="radio"
                value=""
              />
            </span>
            <span
              class="MuiTouchRipple-root-78"
            />
          </span>
          <span
            class="MuiTypography-root-42 MuiTypography-body1-51 MuiFormControlLabel-label-14"
          >
            Private
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root-9"
          id="createNewSharedPipelineBtn"
          title="Everyone in your organization will be able to view and use this pipeline."
        >
          <span
            class="MuiButtonBase-root-30 MuiIconButton-root-24 MuiPrivateSwitchBase-root-20 MuiRadio-root-15 MuiRadio-colorPrimary-18"
          >
            <span
              class="MuiIconButton-label-29"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-33"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
              <input
                class="MuiPrivateSwitchBase-input-23"
                type="radio"
                value=""
              />
            </span>
            <span
              class="MuiTouchRipple-root-78"
            />
          </span>
          <span
            class="MuiTypography-root-42 MuiTypography-body1-51 MuiFormControlLabel-label-14"
          >
            Shared
          </span>
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      Select if the new pipeline will be private or shared.
    </div>
    <div
      class="flex"
    >
      <label
        class="MuiFormControlLabel-root-9"
        id="createNewPrivatePipelineBtn"
        title="Only people who have access to this namespace will be able to view and use this pipeline."
      >
        <span
          class="MuiButtonBase-root-30 MuiIconButton-root-24 MuiPrivateSwitchBase-root-20 MuiRadio-root-15 MuiRadio-colorPrimary-18 MuiPrivateSwitchBase-checked-21 MuiRadio-checked-16"
        >
          <span
            class="MuiIconButton-label-29"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-33"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <input
              checked=""
              class="MuiPrivateSwitchBase-input-23"
              type="radio"
              value=""
            />
          </span>
          <span
            class="MuiTouchRipple-root-78"
          />
        </span>
        <span
          class="MuiTypography-root-42 MuiTypography-body1-51 MuiFormControlLabel-label-14"
        >
          Private
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root-9"
        id="createNewSharedPipelineBtn"
        title="Everyone in your organization will be able to view and use this pipeline."
      >
        <span
          class="MuiButtonBase-root-30 MuiIconButton-root-24 MuiPrivateSwitchBase-root-20 MuiRadio-root-15 MuiRadio-colorPrimary-18"
        >
          <span
            class="MuiIconButton-label-29"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-33"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <input
              class="MuiPrivateSwitchBase-input-23"
              type="radio"
              value=""
            />
          </span>
          <span
            class="MuiTouchRipple-root-78"
          />
        </span>
        <span
          class="MuiTypography-root-42 MuiTypography-body1-51 MuiFormControlLabel-label-14"
        >
          Shared
        </span>
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
